plugins/modules/ec2_vol.py

```diff
@@ -379,125 +385,199 @@ def attach_volume(module, ec2, volume, instance):
-        volume_info['attachment_set']['deleteOnTermination'] = attachment.deleteOnTermination
 
     return volume_info
 
+def get_mapped_block_device(instance_dict=None, device_name=None):
+    mapped_block_device = None
+    if not instance_dict:
+        return mapped_block_device
+    if not device_name:
+        return mapped_block_device
+
+    for device in instance_dict.get('block_device_mappings', []):
+        if device['device_name'] == device_name:
+            mapped_block_device = device
+            break
+
+    return mapped_block_device
+
+
+def ensure_tags(module, connection, res_id, res_type, tags, add_only):
+    changed = False
+
+    filters = ansible_dict_to_boto3_filter_list({'resource-id': res_id, 'resource-type': res_type})
+    cur_tags = None
+    try:
+        cur_tags = connection.describe_tags(Filters=filters)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    purge_tags = bool(not add_only)
+    to_update, to_delete = compare_aws_tags(boto3_tag_list_to_ansible_dict(cur_tags.get('Tags')), tags, purge_tags) #TODO: https://github.com/ansible/ansible/pull/45500/files#r218859482
+    final_tags = boto3_tag_list_to_ansible_dict(cur_tags.get('Tags'))
+
+    if to_update:
+        try:
+            if module.check_mode:
+                # update tags
+                final_tags.update(to_update)
+            else:
+                AWSRetry.exponential_backoff()(connection.create_tags)(
+                    Resources=[res_id],
+                    Tags=ansible_dict_to_boto3_tag_list(to_update)
+                )
+
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't create tags")
+
+    if to_delete:
+        try:
+            if module.check_mode:
+                # update tags
+                for key in to_delete:
+                    del final_tags[key]
+            else:
+                tags_list = []
+                for key in to_delete:
+                    tags_list.append({'Key': key})
+
+                AWSRetry.exponential_backoff()(connection.delete_tags)(Resources=[res_id], Tags=tags_list)
+
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't delete tags")
+
+    if not module.check_mode and (to_update or to_delete):
+        try:
+            response = connection.describe_tags(Filters=filters)
+            final_tags = boto3_tag_list_to_ansible_dict(response.get('Tags'))
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    return final_tags, changed
+
 
 def main():
-    argument_spec = dict(
+    argument_spec = ec2_argument_spec()
```

Comment:
AnsibleAWSModule automatically adds the ec2_argument_spec arguments, this line isn't needed.

Reply:
Thanks for taking a peek @tremble :)

I'm new to `ec2_vol` and Ansible dev in general. I'm still learning how things are being referenced, so if I made any stupid mistakes, please let me know.

I noticed `ec2_argument_spec` returns whatever `aws_common_argument_spec` returns, just with an updated region. I'm thinking the goal here is to then update `ec2_argument_spec`'s dictionary with whatever is passed from module args. Is that right?

Reply:
@tremble ah ok, so we should be good just calling `ec2_argument_spec.update()`, which will update the dictionary that's returned and then pass it to `AnsibleAWSModule`.

Reply:
Even easier than that.

```
argument_spec = dict (.....)
module = AnsibleAWSModule(argument_spec=argument_spec
```

Reply:
The 'update' pieces happen automatically inside AnsibleAWSModule (which in turn wraps AnsibleModule)

Reply:
gotcha :+1:

Reply:
ok, i just pushed the revision up.